File: ai-health-guide-frontend/types/session.ts
/* TypeScript types mirroring the backend SessionState model. */

export type Stage =
  | "intake"
  | "questioning"
  | "visual"
  | "triage"
  | "report"
  | "complete";

export type TriageColor = "RED" | "YELLOW" | "GREEN";

export interface Message {
  role: "patient" | "agent" | "system";
  content: string;
  language: string;
  timestamp: string;
  agent_name?: string | null;
  image_id?: string | null;
}

export interface RedFlag {
  symptom: string;
  severity: "critical" | "high" | "moderate";
  description: string;
  requires_emergency: boolean;
}

export interface ABCDEAssessment {
  airway?: string | null;
  breathing?: string | null;
  circulation?: string | null;
  disability?: string | null;
  exposure?: string | null;
  notes: string;
}

export interface TriageResult {
  color: TriageColor;
  rationale: string;
  facility_type_needed: string;
  urgency_description: string;
  determined_by: "rule_matrix" | "llm_override";
}

export interface PatientReport {
  summary: string;
  what_we_found: string;
  what_to_do_next: string;
  facility_recommendation: string;
  directions_summary: string;
  what_to_tell_doctor: string;
  disclaimer: string;
  language: string;
}

export interface SOAPNote {
  subjective: string;
  objective: string;
  assessment: string;
  plan: string;
  triage_color: TriageColor;
  red_flags_summary: string[];
  language: string;
  patient_language: string;
}

export interface SafetyCheckResult {
  stage: string;
  approved: boolean;
  issues: string[];
  corrective_action: string;
  timestamp: string;
}

export interface SessionState {
  session_id: string;
  created_at: string;
  current_stage: Stage;
  patient_language: string;
  clinical_language: string;
  conversation_history: Message[];

  // Intake
  chief_complaint: string;
  chief_complaint_english: string;

  // Questioning
  structured_symptoms: Record<string, string>;
  abcde_assessment: ABCDEAssessment | null;
  red_flags: RedFlag[];
  severity: number;
  questioning_turns: number;
  questioning_complete: boolean;

  // Visual
  image_data: string | null;
  visual_observations: string[];
  image_analyzed: boolean;

  // Triage
  triage: TriageResult | null;

  // Reports
  patient_report: PatientReport | null;
  clinician_report: SOAPNote | null;

  // Cross-cutting
  safety_checks: SafetyCheckResult[];
  disclaimers_verified: boolean;
}

export interface CreateSessionResponse {
  session_id: string;
}

File: ai_health_guide/models/session.py
from enum import Enum
from datetime import datetime
from typing import Optional, Literal
from uuid import uuid4
from pydantic import BaseModel, Field

from .clinical import ABCDEAssessment, RedFlag, TriageResult
from .report import SOAPNote, PatientReport
from .safety import SafetyCheckResult


class Stage(str, Enum):
    INTAKE      = "intake"
    QUESTIONING = "questioning"
    VISUAL      = "visual"
    TRIAGE      = "triage"
    REPORT      = "report"
    COMPLETE    = "complete"


class Message(BaseModel):
    role:       Literal["patient", "agent", "system"]
    content:    str
    language:   str
    timestamp:  datetime = Field(default_factory=datetime.utcnow)
    agent_name: Optional[str] = None
    image_id:   Optional[str] = None


class Location(BaseModel):
    latitude:  float
    longitude: float
    address:   Optional[str] = None


class Facility(BaseModel):
    name:             str
    place_id:         str
    address:          str
    location:         Location
    facility_type:    str
    distance_meters:  float
    duration_minutes: float
    is_open:          bool
    rating:           Optional[float] = None
    phone:            Optional[str] = None


class SessionState(BaseModel):
    session_id:           str      = Field(default_factory=lambda: str(uuid4()))
    created_at:           datetime = Field(default_factory=datetime.utcnow)
    current_stage:        Stage    = Stage.INTAKE
    patient_language:     str      = "en"
    clinical_language:    str      = "en"
    conversation_history: list[Message] = []

    # Stage 1 — Intake
    chief_complaint:         str = ""
    chief_complaint_english: str = ""

    # Stage 2 — Questioning
    structured_symptoms:  dict                      = {}
    abcde_assessment:     Optional[ABCDEAssessment] = None
    red_flags:            list[RedFlag]             = []
    severity:             int                       = 0
    questioning_turns:    int                       = 0
    questioning_complete: bool                      = False

    # Stage 3 — Visual
    image_data:          Optional[str] = None  # base64
    visual_observations: list[str]     = []
    image_analyzed:      bool          = False

    # Stage 4 — Triage
    triage: Optional[TriageResult] = None

    # Stage 5 — Reports
    patient_report:   Optional[PatientReport] = None
    clinician_report: Optional[SOAPNote]      = None

    # Cross-cutting
    safety_checks:        list[SafetyCheckResult] = []
    disclaimers_verified: bool                    = False
